File: sdks/python/src/pennant/bucketing.py
"""Deterministic bucketing for percentage rollouts.

Mirrors the Pennant server's App\\Services\\Bucketing class byte-for-byte.
The cross-implementation drift corpus at ``tests/corpus/rules.json`` (repo
root) is the contract.
"""

from __future__ import annotations

import hashlib


def bucket(flag_key: str, identifier: str, seed: str) -> float:
    """Return a float in [0, 1) that's deterministic on (flag_key, identifier, seed)."""
    digest = hashlib.sha256(f"{flag_key}:{identifier}:{seed}".encode("utf-8")).hexdigest()
    return int(digest[:8], 16) / 0xFFFFFFFF


def is_in_rollout(flag_key: str, identifier: str, seed: str, percentage: float) -> bool:
    if percentage <= 0:
        return False
    if percentage >= 100:
        return True
    return bucket(flag_key, identifier, seed) < percentage / 100

File: sdks/php/src/Bucketing.php
<?php

declare(strict_types=1);

namespace Pennant;

/**
 * Deterministic bucketing for percentage rollouts.
 *
 * Mirrors the Pennant server's App\Services\Bucketing byte-for-byte; the
 * cross-implementation drift corpus at tests/corpus.json is the contract.
 */
final class Bucketing
{
    public static function bucket(string $flagKey, string $identifier, string $seed): float
    {
        $hex = substr(hash('sha256', $flagKey.':'.$identifier.':'.$seed), 0, 8);
        return ((int) hexdec($hex)) / 0xffffffff;
    }

    public static function isInRollout(string $flagKey, string $identifier, string $seed, float $percentage): bool
    {
        if ($percentage <= 0) return false;
        if ($percentage >= 100) return true;
        return self::bucket($flagKey, $identifier, $seed) < $percentage / 100;
    }
}
